### PR TITLE
Improved graceful shutdown - Collector

### LINKED ISSUE #295: Graceful shutdown by properly handling SIGTERM
From https://github.com/uber/jaeger/pull/280 :

> What I think should happen is you tell the health endpoint to start returning Unavailable and then sleep for a couple seconds, and the exit. The main tchannel / http servers will remain in operation during the sleep and allowed to finish service the in-flight requests, while the routing mesh might hopefully detect health check down and stop sending traffic.

### PR DESCRIPTION
**Pull request by jpkrohling**
## Which problem is this PR solving?
- The Collector parts of #2031 
- See https://github.com/jaegertracing/jaeger/pull/2031#pullrequestreview-359966317 for context

Resolves #295

### CODE REVIEW

**Comment by yurishkuro** [comment]
did you mean to do `git checkout master cmd/agent`?

**Comment by jpkrohling** [comment]
This is on top of #2031. Once that one is merged, this will get rebased and should show only the collector changes.

File: cmd/collector/app/span_processor_test.go [hunk 1]
```diff
@@ -37,8 +37,12 @@ import (
 	zc "github.com/jaegertracing/jaeger/thrift-gen/zipkincore"
 )
 
-var _ (io.Closer) = (*fakeSpanWriter)(nil)
-var blackListedService = "zoidberg"
+var (
+	_ io.Closer = (*fakeSpanWriter)(nil)
+	_ io.Closer = (processor.SpanProcessor)(nil)
```
Comment (annanay25) [inline cmd/collector/app/span_processor_test.go]: `processor.SpanProcessor` is an interface right? That will always pass this test. Should we check `app.spanProcessor`?

File: cmd/collector/app/collector.go [hunk 2]
```diff
@@ -79,57 +82,59 @@ func (c *Collector) Start(builderOpts *CollectorOptions) error {
 		Logger:         c.logger,
 		MetricsFactory: c.metricsFactory,
 	}
-	zipkinSpansHandler, jaegerBatchesHandler, grpcHandler := handlerBuilder.BuildHandlers()
+
+	c.spanProcessor = handlerBuilder.BuildSpanProcessor()
+	c.spanHandlers = handlerBuilder.BuildHandlers(c.spanProcessor)
```
[the remark this answers is not in the document]
Comment (jpkrohling) [reply cmd/collector/app/collector.go]: Good question, probably as part of another PR.

**Comment by jpkrohling** [comment]
@annanay25, @yurishkuro is this ready to be merged once 1.17 is out?

**Comment by yurishkuro** [comment]
I haven't reviewed this one. Let's merge the agent change and rebase this PR.

**Pull request merged**